### ISSUE TITLE
ReqMgr2: cannot update request priority for assignment-approved wfs

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#6961]
Going to reqmgr2 and updating only the `RequestPriority` field raises a 'Server Error' on the browser.
The traceback in the logs is
```
Getting from Cache due to: CouchNotFoundError - reason: Object Not Found, data: {} result: None
[30/Jun/2016:08:47:33]  SERVER REST ERROR WMCore.ReqMgr.DataStructs.RequestError.InvalidSpecParameterValue 9b02ba1fc90401231b5c003a23fe3630 (Invalid spec parameter value: can't update value without state transition: {})
[30/Jun/2016:08:47:33]    Traceback (most recent call last):
[30/Jun/2016:08:47:33]      File "/data/srv/HG1606f-comp4/sw.amaltaro/slc6_amd64_gcc493/cms/reqmgr2/1.0.18.pre3/lib/python2.7/site-packages/WMCore/REST/Server.py", line 708, in default
[30/Jun/2016:08:47:33]        return self._call(RESTArgs(list(args), kwargs))
[30/Jun/2016:08:47:33]      File "/data/srv/HG1606f-comp4/sw.amaltaro/slc6_amd64_gcc493/cms/reqmgr2/1.0.18.pre3/lib/python2.7/site-packages/WMCore/REST/Server.py", line 783, in _call
[30/Jun/2016:08:47:33]        obj = apiobj['call'](*safe.args, **safe.kwargs)
[30/Jun/2016:08:47:33]      File "/data/srv/HG1606f-comp4/sw.amaltaro/slc6_amd64_gcc493/cms/reqmgr2/1.0.18.pre3/lib/python2.7/site-packages/WMCore/ReqMgr/Service/Request.py", line 549, in put
[30/Jun/2016:08:47:33]        result = self._updateRequest(workload, request_args)
[30/Jun/2016:08:47:33]      File "/data/srv/HG1606f-comp4/sw.amaltaro/slc6_amd64_gcc493/cms/reqmgr2/1.0.18.pre3/lib/python2.7/site-packages/WMCore/ReqMgr/Service/Request.py", line 523, in _updateRequest
[30/Jun/2016:08:47:33]        report = self._handleNoStatusUpdate(workload, request_args)
[30/Jun/2016:08:47:33]      File "/data/srv/HG1606f-comp4/sw.amaltaro/slc6_amd64_gcc493/cms/reqmgr2/1.0.18.pre3/lib/python2.7/site-packages/WMCore/ReqMgr/Service/Request.py", line 456, in _handleNoStatusUpdate
[30/Jun/2016:08:47:33]        raise InvalidSpecParameterValue("can't update value without state transition: %s" % request_args)
[30/Jun/2016:08:47:33]    InvalidSpecParameterValue: InvalidSpecParameterValue 9b02ba1fc90401231b5c003a23fe3630 [HTTP 400, APP 1102, MSG "Invalid spec parameter value: can't update value without state transition: {}", INFO None, ERR None]
[30/Jun/2016:08:47:33] alan-cloud1.cern.ch 128.141.136.117 "PUT /reqmgr2/data/request/amaltaro_MonteCarlo_Ext_HG1607_Validation_160629_152252_7641 HTTP/1.1" 400 Bad Request [data: 4695 in 788 out 143966 us ] [auth: OK "/DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=amaltaro/CN=718748/CN=Alan Malta Rodrigues" "" ] [ref: "https://alan-cloud1.cern.ch/reqmgr2/fetch?rid=amaltaro_MonteCarlo_Ext_HG1607_Validation_160629_152252_7641" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:47.0) Gecko/20100101 Firefox/47.0" ]
```

Ideally, all parameters *not* changed should be removed from the json passed upstream.
This goes inline with this issue as well #6932

**Comment by amaltaro** [comment]
@vkuznet how easy is it to pass only modified parameters upstream? Ideally we should make this change for ALL pages where there is a possible action.
If possible, we could target it for the next cmsweb release (the current one is too tight at this moment)

**Comment by vkuznet** [comment]
Alan, I don't see relation of posted traceback with your request feature. The traceback just says that it can't update state without state transition. The web UI should not make decision which parameters are accepted or not on a backend server it is a server job. The web UI should parse the web form that's it and parsing the web form means take all parameters on it. Don't delegate the action to inappropriate layer. If server need to do a PUT request with the same parameters nothing wrong with it, think about it. You have a doc with some values and you want to update it. If you'll be given the same parameters to update nothing wrong with that.

**Comment by amaltaro** [comment]
The traceback is misleading because the web UI is passes the whole json back to the server, regardless of what changed or not. Which then the backend server handles as a "whole" request update instead of a requestpriority only (there are different methods on the server to handle that).

I _may_ agree with you that such logic should not be put on the web layer. However, since it has all the values already why to verify what is new and only pass those parameters?
Otherwise, we need to make another GET call on the server for every PUT call, to fetch the data and x-check what is updated or not. Do you see my point?

**Comment by ticoann** [comment]
There is a restriction on assignment-approved to assigned transition. 
Properties cannot be updated without state transition.
If that is not reasonable assumption, I can can change that in serverside.

**Comment by amaltaro** [comment]
If that's not too much work, I think allowing RequestPriority change at any request status before completed should be done. A use case would be: McM injects a small batch and notices their priority is wrong, so they'd simply update it without state transition.

**Comment by amaltaro** [comment]
I just gave it a try via web UI on request:
amaltaro_TaskChain_MC_HG1609_ReqMgr2_Validation_160906_130245_5036

and got back a "Server error message". The traceback is:
```
Getting from Cache due to: CouchNotFoundError - reason: Object Not Found, data: {} result: None
[06/Sep/2016:13:52:38]  Error: Traceback (most recent call last):
  File "/data/srv/beHG1609c/sw/slc6_amd64_gcc493/cms/reqmgr2/1.0.20.pre4/lib/python2.7/site-packages/WMCore/ReqMgr/Service/Request.py", line 205, in validate
    self._validateRequestBase(param, safe, validate_request_update_args, requestName)
  File "/data/srv/beHG1609c/sw/slc6_amd64_gcc493/cms/reqmgr2/1.0.20.pre4/lib/python2.7/site-packages/WMCore/ReqMgr/Service/Request.py", line 113, in _validateRequestBase
    workload, r_args = valFunc(args, self.config, self.reqmgr_db_service, param)
  File "/data/srv/beHG1609c/sw/slc6_amd64_gcc493/cms/reqmgr2/1.0.20.pre4/lib/python2.7/site-packages/WMCore/ReqMgr/Utils/Validation.py", line 116, in validate_request_update_args
    workload.validateArgumentForAssignment(args_without_status)
  File "/data/srv/beHG1609c/sw/slc6_amd64_gcc493/cms/reqmgr2/1.0.20.pre4/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkload.py", line 1650, in validateArgumentForAssignment
    validateArgumentsUpdate(schema, argumentDefinition)
  File "/data/srv/beHG1609c/sw/slc6_amd64_gcc493/cms/reqmgr2/1.0.20.pre4/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkloadTools.py", line 250, in validateArgumentsUpdate
    _validateArgumentOptions(arguments, argumentDefinition, "assign_optional")
  File "/data/srv/beHG1609c/sw/slc6_amd64_gcc493/cms/reqmgr2/1.0.20.pre4/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkloadTools.py", line 124, in _validateArgumentOptions
    raise WMSpecFactoryException(msg)
WMSpecFactoryException: WMSpecFactoryException
Message: Validation failed: ProcessingString is mandatory {'attr': 'processingString', 'default': '', 'assign_optional': False, 'optional': True, 'validate': <function procstring at 0x7f036ff35e60>, 'null': False, 'type': <type 'str'>}
	ModuleName : WMCore.WMSpec.WMWorkloadTools
	MethodName : _validateArgumentOptions
	ClassInstance : None
	FileName : /data/srv/beHG1609c/sw/slc6_amd64_gcc493/cms/reqmgr2/1.0.20.pre4/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkloadTools.py
	ClassName : None
	LineNumber : 124
	ErrorNr : 0

Traceback: 



[06/Sep/2016:13:52:38]  SERVER REST ERROR WMCore.ReqMgr.DataStructs.RequestError.InvalidSpecParameterValue a56ef5fdc300f66feb59c298f0b0af26 (Invalid spec parameter value: Validation failed: ProcessingString is mandatory {'attr': 'processingString', 'default': '', 'assign_optional': False, 'optional': True, 'validate': <function procstring at 0x7f036ff35e60>, 'null': False, 'type': <type 'str'>})
[06/Sep/2016:13:52:38]    Traceback (most recent call last):
[06/Sep/2016:13:52:38]      File "/data/srv/beHG1609c/sw/slc6_amd64_gcc493/cms/reqmgr2/1.0.20.pre4/lib/python2.7/site-packages/WMCore/REST/Server.py", line 708, in default
[06/Sep/2016:13:52:38]        return self._call(RESTArgs(list(args), kwargs))
[06/Sep/2016:13:52:38]      File "/data/srv/beHG1609c/sw/slc6_amd64_gcc493/cms/reqmgr2/1.0.20.pre4/lib/python2.7/site-packages/WMCore/REST/Server.py", line 779, in _call
[06/Sep/2016:13:52:38]        v(apiobj, request.method, api, param, safe)
[06/Sep/2016:13:52:38]      File "/data/srv/beHG1609c/sw/slc6_amd64_gcc493/cms/reqmgr2/1.0.20.pre4/lib/python2.7/site-packages/WMCore/ReqMgr/Service/Request.py", line 238, in validate
[06/Sep/2016:13:52:38]        raise InvalidSpecParameterValue(msg)
[06/Sep/2016:13:52:38]    InvalidSpecParameterValue: InvalidSpecParameterValue a56ef5fdc300f66feb59c298f0b0af26 [HTTP 400, APP 1102, MSG "Invalid spec parameter value: Validation failed: ProcessingString is mandatory {'attr': 'processingString', 'default': '', 'assign_optional': False, 'optional': True, 'validate': <function procstring at 0x7f036ff35e60>, 'null': False, 'type': <type 'str'>}", INFO None, ERR None]
```

**Comment by ticoann** [comment]
Alan, it won't allow only Priority update via Web interface. (It should allow via REST api).
The problem with GUI is that it will send all the data in the page when it updates the priority, which is not allowed. 
What we can do is when there is no state transition it only update the Priority. But IMO, it is a little confusing behavior. i.e if someone changes 2 field including Priority and other field but it only gets updated one.

**Comment by amaltaro** [comment]
> What we can do is when there is no state transition it only update the Priority. But IMO, it is a little confusing behavior. i.e if someone changes 2 field including Priority and other field but it only gets updated one.

That's exactly what we should do, IMO. Either:
a) Ops change the RequestPriority only (no need to validate/update anything else)
b) Ops change RequestStatus and update all the other assignment parameters
c) same as b), but Ops also updates the RequestPriority.

Anything that doesn't meet these use cases, should fail.

**Comment by amaltaro** [comment]
Fixed by #7265
